File: app/src/reconciliation/reconciler.js
const logger = require('../utils/logger');
const retryQueue = require('../utils/retryQueue');
const { checkPackageDelivered, insertInvoice } = require('../ingestion/eventHandlers');

async function runReconciliation() {
  const queue = retryQueue.getAll();

  if (queue.length === 0) {
    logger.info('Retry queue is empty, nothing to reconcile');
    return { processed: 0, remaining: 0 };
  }

  logger.info(`Starting reconciliation for ${queue.length} deferred events`);

  let processedCount = 0;

  for (const event of queue) {
    const { invoice_id, package_id, amount, customer_id, currency, timestamp, event_id } = event;

    const isDelivered = await checkPackageDelivered(package_id);

    if (!isDelivered) {
      logger.info('Package still not DELIVERED, keeping in retry queue', {
        package_id,
        invoice_id,
      });
      continue;
    }

    logger.info('Package is now DELIVERED, processing deferred billing event', {
      package_id,
      invoice_id,
    });

    try {
      await insertInvoice({ invoice_id, package_id, amount, customer_id, currency, timestamp, event_id });
      retryQueue.dequeue(invoice_id);
      processedCount++;
    } catch (err) {
      logger.error('Failed to process deferred billing event during reconciliation', {
        invoice_id,
        package_id,
        error: err.message,
      });
    }
  }

  const remaining = retryQueue.getAll().length;
  logger.info('Reconciliation complete', { processed: processedCount, remaining });
  return { processed: processedCount, remaining };
}

module.exports = { runReconciliation };

File: app/src/reconciliation/reconciler.py
from src.utils import retry_queue
from src.utils.logger import get_logger
from src.ingestion.event_handlers import check_package_delivered, insert_invoice

logger = get_logger(__name__)


def run_reconciliation() -> dict:
    queue = retry_queue.get_all()

    if not queue:
        logger.info("Retry queue is empty, nothing to reconcile")
        return {"processed": 0, "remaining": 0}

    logger.info(f"Starting reconciliation for {len(queue)} deferred events")
    processed_count = 0

    for event in queue:
        invoice_id = event.get("invoice_id")
        package_id = event.get("package_id")

        is_delivered = check_package_delivered(package_id)

        if not is_delivered:
            logger.info(f"Package still not DELIVERED, keeping in retry queue: package_id={package_id} invoice_id={invoice_id}")
            continue

        logger.info(f"Package is now DELIVERED, processing deferred billing event: package_id={package_id} invoice_id={invoice_id}")

        try:
            insert_invoice(event)
            retry_queue.dequeue(invoice_id)
            processed_count += 1
        except Exception as e:
            logger.error(f"Failed to process deferred billing event: invoice_id={invoice_id} error={e}")

    remaining = len(retry_queue.get_all())
    logger.info(f"Reconciliation complete: processed={processed_count} remaining={remaining}")
    return {"processed": processed_count, "remaining": remaining}
